### PR TITLE
Add interpolation

### PR DESCRIPTION
**Pull request by AbelVM**
- [x] All declared geometries are `geometry(Geometry, 4326)` for general geoms, or `geometry(Point, 4326)`
- [x] Docs for public-facing functions are written
- [x] New functions follow the naming conventions: `CDB_NameOfFunction`. Where internal functions begin with an underscore `_`.
- [x] If appropriate, new functions accepts an arbitrary query as an input (see [Crankshaft Issue #6](https://github.com/CartoDB/crankshaft/issues/6) for more information)

### CODE REVIEW

File: doc/08_interpolation.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+## Spacial interpolation
+
+Function to interpolate a numeric attribute of a point in a scatter dataset of points, using one of three methos:
+
+* [Nearest neighbor](https://en.wikipedia.org/wiki/Nearest-neighbor_interpolation)
+* [Barycentric](https://en.wikipedia.org/wiki/Barycentric_coordinate_system)
+* [IDW](https://en.wikipedia.org/wiki/Inverse_distance_weighting)
+
+### CDB_SpatialInterpolation (query text, point geometry, method integer DEFAULT 1, p1 integer DEFAULT 0, ps integer DEFAULT 0)
```
Comment (javisantana) [inline doc/08_interpolation.md]: I'd remove this signature, having a query as an argument but looks likes this was a research request cc @andrewxhill

**Comment by andrewxhill** [comment]
cc @stuartlynn ^ see santana's comment

**Comment by stuartlynn** [comment]
Happy to drop it for this function and happy to find a better way of doing this in general but we have tried a lot here to think of one without much success. Let us know if we are missing something or you have a better way of doing this with plpythonu

**Comment by javisantana** [comment]
some example on how to do aggregates https://github.com/CartoDB/cartodb-postgresql/blob/master/scripts-available/CDB_Math.sql

Anyway, if we follow the postgis way to do this we should take a look at `ST_Line_Interpolate_Point`. That function interpolates taking into account the Z and M coordinates.

We could have:

```
select CDB_SpatialInterpolation(st_makepoint(st_x(the_geom), st_y(the_geom), value) from table limit 400 
```

with 

````
CREATE AGGREGATE CDB_SpatialInterpolation (
  BASETYPE = geometry,
  SFUNC = accum_fn,
  STYPE = TYPE
);
```

My 2 cents

**Comment by stuartlynn** [comment]
Yeah we know how to do aggregates but what about functions that require us to act on the dataset as a whole, like the pysal's morans I or a machine learning algorithm for example? Unless we want to write all those methods from scratch in SQL, which would be hugely inefficient and require a lot of code, then aggregates don't help. 

We would basically be generating an aggregate to grab all the data and then in the ffunc phase pass it to the python function anyway which seems just as efficient or less so that using plpy.execute to grab the data from the table.

Or for example with machine learning, if we train the model row by row in an aggregate, persisting state between each aggregate call its going to be massively slow as it cant take advantage of the linear algebra libraries that underpin the python methods. 

You see a similar approach to the query passing by string in MadLib for example here : http://madlib.incubator.apache.org/docs/latest/group__grp__decision__tree.html which is the closest thing I have seen to what we are trying to do in crankshaft. They pass it in as parts but they must be doing something similar to under the hood.

**Comment by javisantana** [comment]
There are several ways to do those things, window functions for example: https://gist.github.com/dbaston/6f76c3462e40a69518cd

Anyway, let's get to moran's, let me ask some questions, as I understand Morans is doing actually more than one thing, first [looking for neighbors ](https://github.com/CartoDB/crankshaft/blob/develop/src/py/crankshaft/crankshaft/pysal_utils/pysal_utils.py#L75) and then doing some magic work so:

- why we don't have a function that returns those neighbors in pure SQL ? You are kind of doing that in sql from python.
- Is the python query+aggregation faster than the SQL query + aggregation ? Because as i see we could do the neighbors query and then pass that result to morans algortihm 
- Is the python function using any index after the neighbors search?

**Comment by javisantana** [comment]
also, FYI:

```sql
CREATE OR REPLACE FUNCTION py_array(x integer[])
RETURNS float
AS $$
    return sum(x)
$$ LANGUAGE plpythonu;

CREATE OR REPLACE FUNCTION py_query()
RETURNS text
AS $$
    import plpy
    res = plpy.execute('SELECT x from generate_series(1, 1000000) x');
    return sum(map(lambda x: x['x'], res))
$$ LANGUAGE plpythonu;


\timing
SELECT py_array(array_agg(x)) from generate_series(1, 1000000) x;
select py_query();
```

the execution:

```sql
psql:/Users/j/tmp/test.sql:19: LOG:  statement: SELECT py_array(array_agg(x)) from generate_series(1, 1000000) x;
   py_array
--------------
 500000500000
(1 row)

Time: 347.754 ms
psql:/Users/j/tmp/test.sql:20: LOG:  statement: select py_query();
   py_query
--------------
 500000500000
(1 row)

Time: 945.630 ms
```

**Comment by stuartlynn** [comment]
I agree windows functions would be great but the API is only available for C and PL/v8 not plpsql or plpythonu. We could move all of the analysis over to C but then we have to reimplement a huge amount of code that is already available and well tried and tested in the python libraries we have been using. 

http://stackoverflow.com/questions/13790028/how-to-create-a-custom-windowing-function-for-postgresql-running-average-examp

For Morans yeah we could take what @andy has done and make it pure SQL for the first step that would be good. We are trying to do as much in pure SQL but that doesn't side step the problem, we still at some point need a chunk of data in python to process.

The index is most important in the spatial query parts but once we have identified the data then no python doesn't use the index but it does use a lot of internal optimization bellow the hood to do the more linear algebra heavy parts of the computation.

**Comment by stuartlynn** [comment]
With the timing example you posted most of the slow down is in python deserializing the object that plpy.execute returns. I have been playing to find ways to make this faster for example 

```postgresql 
CREATE OR REPLACE FUNCTION py_query2()
RETURNS text AS $$
import plpy
res = plpy.execute('SELECT array_agg(x) as x from generate_series(1, 1000000) x');
return sum(res[0]['x'])
$$ LANGUAGE plpythonu;
```
```postgresql
select py_query2();                                                                                                                                                                                                      py_query2   
--------------
 500000500000
(1 row)

Time: 303.743 ms
```

For reference on my machine your py_array function ran in 343.433ms and py_query in 999.518 

So I don't think that the inefficiency is coming from calling the the query inside or outside of plpython but rather in the way plpy.execute returns its query object.

**Comment by javisantana** [comment]
that deserialization is what moran function is doing, that's why I did it in that way.

The point here is, there is no difference between doing outside and inside in terms of performance but the second way makes everything to be the same as other libraries (postgres and postgis mostly).

But, there are other stuff we should think about. A function that accepts a query is a smell of "this should run outside of the database"

I totally understand we are doing it because of the time we have but please, do not confuse this with doing things in the right way.

**Comment by stuartlynn** [comment]
I dont like this way of doing things either and I definitely wouldn't say its the ideal way but we have spent a significant chunk of time of the past few months trying to find alternatives that work for the use cases we have. Where we have found them we have implemented them, like @ohasselblad implementing the spatial weights function in SQL or starting to use that array aggregation trick to speed up de-serialisation (the reason it isn't on the Moran functions is cause we have been swamped with other things and haven't had time to refactor it). 

We aren't doing things this way because we want to it's just that after a long time thinking about it we have been unable to find alternatives. There are real, very annoying constraints with using plpythonu like the inability to pass multidimensional arrays, fixed function signatures (not overloading but being able to have a variable number of inputs. For example with ML this is important) and the plpy.execute return form. We are trying to work around them as best we can.

**Comment by ohasselblad** [comment]
Ok, we're not going to use the version of the function that accepts the query, but will instead use the one that uses this signature:

```
CREATE OR REPLACE FUNCTION CDB_SpatialInterpolation(
    IN geomin geometry[],
    IN colin numeric[],
    IN point geometry,
    IN method integer DEFAULT 1,
    IN p1 numeric DEFAULT 0,
    IN p2 numeric DEFAULT 0
    )
RETURNS numeric AS
```

The camshaft node I'll write be something like this:
```
var weightedKnnQueryTemplate = dot.template([
    'SELECT cdb_crankshaft.CDB_SpatialInterpolation(',
    '  array_agg(the_geom), ',
    '  array_agg({{=it.val_column}}), ',
    '  {{=it.source_geom}}, ',
    '  {{=it.num_neighbors}}',
    '  2') As knn_averaged ',
    'FROM ({{=it.query}}) As q'
  ].join('\n'));
```

**Comment by ohasselblad** [comment]
@javisantana, ^^ ok with you?

**Comment by javisantana** [comment]
sounds good, thanks!

**Comment by AbelVM** [comment]
Hmm...  but there should be nodes for the 3 methods and with all the parameters free... no?

I mean, IDW is only usefull for some analysis and fixing the distance decay to 2nd order may or may not be always the right option.

verbi Gratia:
For fast analisys: nearest neighbor
For Z value: barymetric

**Comment by ohasselblad** [comment]
We were thinking of exposing them for different analyses.. the first one i'd tackle is just the interpolated nearest neighbors with decay 1/d^2. @AbelVM, I'll cc you in the camshaft pr so we can have discuss there.

**Pull request merged**